src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Extensibility/ProjectExportProvider.cs

```diff
@@ -0,0 +1,84 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.Composition;
+using System.Linq;
+using Microsoft.VisualStudio.Composition;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Extensibility
+{
+    /// <summary>
+    /// MEF component which has methods for consumers to get to project specific MEF exports
+    /// </summary>
+    [Export(typeof(IProjectExportProvider))]
+    [AppliesTo(ProjectCapability.CSharpOrVisualBasic)]
+    internal class ProjectExportProvider : IProjectExportProvider
+    {
+        private SVsServiceProvider ServiceProvider { get; set; }
+
+        [ImportingConstructor]
+        public ProjectExportProvider(SVsServiceProvider serviceProvider)
+        {
+            ServiceProvider = serviceProvider;
+        }
+
+        /// <summary>
+        /// This function requires going to the UI thread to map the hierarchy to the correct project
+        /// instance. 
+        /// </summary>
+        public T GetExport<T>(IVsHierarchy projectHierarchy) where T : class
+        {
+            if (projectHierarchy == null)
+            {
+                throw new ArgumentNullException(nameof(projectHierarchy));
+            }
+
+            T exportedValue = null;
+
+            ThreadHelper.JoinableTaskFactory.Run(async () =>
+            {
+                await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
```

Comment:
The above construct seems like it will result in a deadlock. Please correct me if I am wrong.

* The calling thread for `GetExport` is already in UI thread
* `ThreadHelper.JoinableTaskFactory.Run` blocks the UI thread to run the lambda
*  'ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync' waits for the UI thread to be free

Hence deadlock. Am I missing something here?